crypto/o_fips.c

```diff
@@ -1,24 +1,62 @@
 /*
- * Copyright 2011-2016 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2011-2020 The OpenSSL Project Authors. All Rights Reserved.
  *
  * Licensed under the Apache License 2.0 (the "License").  You may not use
  * this file except in compliance with the License.  You can obtain a copy
  * in the file LICENSE in the source distribution or at
  * https://www.openssl.org/source/license.html
  */
 
+#include "crypto/evp.h"
 #include "internal/cryptlib.h"
 
+#define FIPS_QUERY_ON  "fips=yes"
+#define FIPS_QUERY_OFF ""
+#define DIGEST_NAME "SHA-256"
+
+/*
+ * Check that the library context's default properties has the value "fips=yes"
+ */
+int FIPS_mode_with_libctx(OPENSSL_CTX *libctx)
+{
+    return evp_default_property_is_enabled(libctx, "fips");
+}
+
 int FIPS_mode(void)
 {
-    /* This version of the library does not support FIPS mode. */
-    return 0;
+    return FIPS_mode_with_libctx(NULL);
+}
+
+/*
+ * Sets the default properties for a library context to either
+ * “fips=yes” if on != 0 OR
+ * “” for on == 0. i.e (fips is unset)
+ * If on != 0 it also fetches the SHA256 algorithm with the property
+ * “fips=yes” and ensure that it successfully returns.
+ */
+int FIPS_mode_set_with_libctx(OPENSSL_CTX *libctx, int on)
+{
+    int ret = 1;
+    int last_on = FIPS_mode_with_libctx(libctx);
+
+    if (!EVP_set_default_properties(libctx, on != 0 ? FIPS_QUERY_ON :
```

Comment:
A thought.  Instead of completely replacing the global property.  Would a merge be nicer.  For turning FIPS on, merge "fips=yes" over the default properties.  For turning off, merge "-fips".  The low level `ossl_property_merge` and `ossl_property_match_count` functions know what to do.

Just a thought.

Reply:
More specifically, replace `ossl_method_store_set_global_properties' with `ossl_method_store_merge_global_properties` defined like:
```C
int ossl_method_store_merge_global_properties(OSSL_METHOD_STORE *store,
                                              const char *prop_query) {
    int ret = 0;
    OSSL_PROPERTY_LIST *prop = NULL, *global;

    if (store == NULL)
        return 1;

    ossl_property_write_lock(store);
    ossl_method_cache_flush_all(store);
    if (prop_query == NULL) {
        ossl_property_free(store->global_properties);
        store->global_properties = NULL;
        goto success;
    }
    prop = ossl_parse_query(store->ctx, prop_query);
    if (prop == NULL)
        goto end;

    if (store->global_properties == NULL) {
        store->global_properties = prop;
        prop = NULL;
        goto success;
    }
    global = ossl_property_merge(prop, store->global_properties);
    if (global == NULL)
        goto end;
    ossl_property_free(store->global_properties);
    store->global_properties = global;
 success:
    ret = 1;
 end:
    ossl_property_unlock(store);
    ossl_property_free(prop);
    return ret;
}
```

Reply:
-fips would not work probably? Because the serializers have fips=yes but you do not want to turn them off.

Reply:
I have added this function in - thanks Pauli